File: src/main/java/dev/jlo/gamemodes/domain/common/MatchRoster.java
package dev.jlo.gamemodes.domain.common;

import java.util.EnumMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

public final class MatchRoster {
    private final int teamCapacity;
    private final Quorum quorum;
    private final Map<Team, LinkedHashSet<UUID>> members = new EnumMap<>(Team.class);

    public MatchRoster(int teamCapacity, Quorum quorum) {
        if (teamCapacity <= 0) {
            throw new IllegalArgumentException("Team capacity must be positive");
        }
        this.teamCapacity = teamCapacity;
        this.quorum = quorum;
        members.put(Team.A, new LinkedHashSet<>());
        members.put(Team.B, new LinkedHashSet<>());
    }

    public int getTeamCapacity() {
        return teamCapacity;
    }

    public Quorum getQuorum() {
        return quorum;
    }

    public Team join(UUID player) {
        if (members.values().stream().anyMatch(team -> team.contains(player))) {
            throw new IllegalStateException("Player is already in this roster");
        }
        Team selected = Team.A;
        for (Team team : Team.values()) {
            if (members.get(team).size() < members.get(selected).size()) {
                selected = team;
            }
        }
        if (members.get(selected).size() >= teamCapacity) {
            throw new IllegalStateException("Both teams are full");
        }
        members.get(selected).add(player);
        return selected;
    }

    public Team leave(UUID player) {
        for (Map.Entry<Team, LinkedHashSet<UUID>> entry : members.entrySet()) {
            if (entry.getValue().remove(player)) {
                return entry.getKey();
            }
        }
        return null;
    }

    public Set<UUID> players(Team team) {
        return Set.copyOf(members.get(team));
    }

    public Team teamOf(UUID player) {
        for (Map.Entry<Team, LinkedHashSet<UUID>> entry : members.entrySet()) {
            if (entry.getValue().contains(player)) {
                return entry.getKey();
            }
        }
        return null;
    }

    public boolean hasQuorum() {
        for (Team team : Team.values()) {
            if (members.get(team).size() < quorum.getRequiredPlayers()) {
                return false;
            }
        }
        return true;
    }
}

File: src/main/kotlin/dev/jlo/gamemodes/domain/common/MatchRoster.kt
package dev.jlo.gamemodes.domain.common

import java.util.UUID
import kotlin.math.ceil

enum class Team {
    A,
    B
}

data class Quorum private constructor(val requiredPlayers: Int) {
    init {
        require(requiredPlayers > 0) { "Quorum must require at least one player" }
    }

    companion object {
        fun fixed(requiredPlayers: Int): Quorum = Quorum(requiredPlayers)

        fun minimumPlayers(capacityPerTeam: Int, fraction: Double): Quorum {
            require(capacityPerTeam > 0) { "Capacity must be positive" }
            require(fraction > 0.0 && fraction <= 1.0) { "Quorum fraction must be in (0, 1]" }
            return Quorum(ceil(capacityPerTeam * fraction).toInt())
        }
    }
}

class MatchRoster(private val teamCapacity: Int, private val quorum: Quorum) {
    private val members = linkedMapOf<Team, LinkedHashSet<UUID>>(
        Team.A to linkedSetOf(),
        Team.B to linkedSetOf()
    )

    init {
        require(teamCapacity > 0) { "Team capacity must be positive" }
    }

    fun join(player: UUID): Team {
        check(members.values.none { player in it }) { "Player is already in this roster" }
        val team = members.minWith(compareBy<Map.Entry<Team, LinkedHashSet<UUID>>> { it.value.size }.thenBy { it.key.ordinal }).key
        check(members.getValue(team).size < teamCapacity) { "Both teams are full" }
        members.getValue(team).add(player)
        return team
    }

    fun leave(player: UUID): Team? = members.entries.firstOrNull { it.value.remove(player) }?.key

    fun players(team: Team): Set<UUID> = members.getValue(team).toSet()

    fun teamOf(player: UUID): Team? = members.entries.firstOrNull { player in it.value }?.key

    fun hasQuorum(): Boolean = Team.entries.all { members.getValue(it).size >= quorum.requiredPlayers }
}
